extra/crc32-x86-pclmul/crc_intel_pclmul.c

```diff
@@ -508,4 +510,41 @@ crc32_intel_pclmul (u32 *pcrc, const byte *inbuf, size_t inlen)
 #endif
 }
 
+static inline uint32_t cpuid(uint32_t *ecx, uint32_t *edx)
+{
+  uint32_t level;
+
+  asm("cpuid" : "=a"(level) : "a"(0) : "ebx", "ecx", "edx");
+
+  if (level < 1)
+  {
+    return level;
+  }
+
+  asm("cpuid" : "=c"(*ecx), "=d"(*edx) : "a"(1) : "ebx");
+
+  return level;
+}
+
+int is_pclmul_enabled()
+{
+  int pclmul_enabled= 0;
+  uint32_t ecx, edx;
+
+  if (cpuid(&ecx, &edx) > 0)
+  {
+    pclmul_enabled= ((ecx >> 19) & 1) && ((ecx >> 1) & 1);
+  }
+
+  return (pclmul_enabled);
+}
```

Comment:
Motivated by @grooverdan’s suggestions I made some experiments and came up with the following:
```c
#ifdef __GNUC__
int has_crc32c() { return __builtin_cpu_supports("sse4.2"); }
# if __GNUC__ >= 6 || defined __clang__
int has_pclmul() { return __builtin_cpu_supports("pclmul"); }
# else
int has_pclmul() { return 0; }
# endif
#elif defined _MSC_VER
#include <intrin.h>
int has_crc32c()
{
    int regs[4];
    __cpuid(regs, 1);
    return regs[2] & 1 << 20;
}
int has_pclmul()
{
    int regs[4];
    __cpuid(regs, 0);
    return regs[2] & 1 << 19;
}
#endif
```

For some reason, `clang 4.0.0` and GCC 6.1 (and later versions that I checked on https://godbolt.org/) only check the `ecx & (1 << 19)` bit:
```asm
has_pclmul():                        # @has_pclmul()
        movl    __cpu_model+12(%rip), %eax
        shrl    $19, %eax
        andl    $1, %eax
        retq
```
We will only need the `has_pclmul()` here. The `has_crc32c()` would replace the need for `ut_crc32()` call in InnoDB.